### PR TITLE
ARROW-420: Align DATE type with Java implementation

### PR DESCRIPTION
**Pull request by xhochy**

### CODE REVIEW

**Comment by xhochy** [comment]
Things to do: 

- [ ] Update parquet-cpp
- [ ] Add conversion from `datetime.date` to `DateArray`
- [ ] Add python unit test

**Comment by wesm** [comment]
Why would we need more than `int32_t` for date?

**Comment by xhochy** [comment]
See the ticket https://issues.apache.org/jira/browse/ARROW-413

The way DATE is implemented in Java is: "Here's a millisecond timestamp but please only look at the days".

**Comment by xhochy** [comment]
Ready for review. As `parquet-cpp` builts without changes, I'd first like to get this here merged and would then add in another iteration support for `DATE` in `parquet-cpp`.

File: python/src/pyarrow/adapters/pandas.cc [hunk 1]
```diff
@@ -369,6 +392,10 @@ inline Status ArrowSerializer<NPY_OBJECT>::Convert(std::shared_ptr<Array>* out)
 
   // TODO: mask not supported here
   const PyObject** objects = reinterpret_cast<const PyObject**>(PyArray_DATA(arr_));
+  {
+    PyAcquireGIL lock;
+    PyDateTime_IMPORT;
```
Comment (wesm) [inline python/src/pyarrow/adapters/pandas.cc]: You're right, we should acquire the GIL below before using any of the Py* methods. I'll create a JIRA about going through this module completely and making sure we're releasing the GIL on the Cython side and re-acquiring it in the appropriate places here

**Review by wesm — APPROVED** [review]
+1, creating follow up JIRA about GIL questions

**Pull request closed without merging**